provisioning/containers/gui-common/websockify/instmetrics.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2020-2022 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"math"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// Resources utilization derived from ContainerMetrics.
+type Resources struct {
+	Connections []ConnInfo `json:"connections"`
+	CPUPerc     uint16     `json:"cpu"`
+	MEMPerc     uint16     `json:"mem"`
+	NET         int64      `json:"net"`
+	Timestamp   time.Time  `json:"timestamp,omitempty"`
+	ErrorMsg    string     `json:"error,omitempty"`
+}
+
+// InstanceMetricsHandler is the main handler for instance metrics.
+type InstanceMetricsHandler struct {
+	instanceMetricsClient *InstanceMetricsClient
+	// Application container resources.limits.cpu.
+	cpuLimit string
+	// Application container resources.limits.memory.
+	memoryLimit string
+	// PodName where application container is running.
+	podName string
+	// Last Resources extracted from CustomMetricsServer
+	cachedResources      *Resources
+	cachedResourcesMutex sync.RWMutex
+	// <UID, ConnInfo> map
+	connectionsTracking *sync.Map
+}
+
+// ServeHTTP serves WS server.
+func (h *InstanceMetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
```

Comment:
Would do all the various parameters check and retrieval before upgrading the connection.